## Задача 3.3
Поиск параметра формы $\gamma$ в распределении Вейбулла

In [106]:
%matplotlib inline  
import scipy.stats as st
import numpy as np
import matplotlib.pyplot as plt
import math
import statsmodels.api as sm
import csv

Так как в выборке встречаются нули, а их значение из выборки принимается с нулевой вероятностью, то 
заменим нули выборки на минимальное не нулевое значение выборки

In [107]:
from numpy import genfromtxt
my_data = genfromtxt('Weibull.csv', delimiter=',')
minimum = np.amin(my_data[my_data > 0])
my_data[my_data == 0] = minimum

Введем функцию вычисляющую логарифмическую функцию правдоподобия для одного наблюдения

In [108]:
def my_func(x, t):
    return st.exponweib.logpdf(x, 1, t, loc=0, scale=1)

### Оцениваем параметр на выборке по 4 годам

Будем считать по логарифмической сетке $[-2, 2]$ с шагом $0.001$ на выборке по 4 годам

In [109]:
max_res=-math.inf
max_i = 0

for i in np.arange(-2, 2, 0.001):
    results = np.sum(my_func(my_data[:1460], 10**i))
    if results > max_res:
        max_res = results
        max_i = i
print("Максимум логарифмической функции правдоподобия: " + str(max_res))
print("На каком параметре достигается максимум на логарифмической сетке: " + str(max_i))
print("Оценка парамета гамма: " + str(10**max_i))

Максимум логарифмической функции правдоподобия: -2025.9740355717613
На каком параметре достигается максимум на логарифмической сетке: -0.44700000000017104
Оценка парамета гамма: 0.3572728381517882


В результате получили оценку параметра $\gamma = 0.357$ на которой достигается максимум функции правдоподобия

### Оцениваем параметр по всей выборке

Будем считать по логарифмической сетке $[-2, 2]$ с шагом $0.001$ на всей выборке

In [110]:
max_res=-math.inf
max_i = 0

for i in np.arange(-2, 2, 0.001):
    results = np.sum(my_func(my_data, 10**i))
    if results > max_res:
        max_res = results
        max_i = i
print("Максимум логарифмической функции правдоподобия: " + str(max_res))
print("На каком параметре достигается максимум на логарифмической сетке: " + str(max_i))
print("Оценка парамета гамма: " + str(10**max_i))

Максимум логарифмической функции правдоподобия: -5056.799964821297
На каком параметре достигается максимум на логарифмической сетке: -0.4520000000001705
Оценка парамета гамма: 0.3531831697918183


В результате получили оценку параметра $\gamma = 0.353$ на которой достигается максимум функции правдоподобия

### Вывод
С помощью метода максимального правдоподобия был оценен параметр формы для распределения Вейбулла.
На 4-летней и полной выборке получился приблизительно одиннаковый результат.